Fixed #18468 -- Added support for comments on columns and tables.

KimSoungRyoul opened the pull request:
more detail description: https://code.djangoproject.com/ticket/18468


#### Django
* Field(db_comment="...")
* db_table_comment="...."

~~~Python
class AModel(Model):
  
  a_char = models.Field(db_comment="I Am comment to Column", max_length=32)
  
  class Meta:
      db_table = "a_model"
      db_table_comment = "table comment...."

~~~


#### Converted to postgres dialect

~~~SQL

create table a_model
(
    id      bigserial   not null  constraint hello1_amodel_pkey  primary key,
    a_char  varchar(32) not null
    
);

comment on table a_model is 'table comment....';

comment on column a_model.a_char is 'I Am comment to Column';
~~~


#### Converted to mysql  dialect

~~~SQL
create table a_model33
(
    id     bigint auto_increment primary key,
    a_char varchar(32) not null comment 'I AM Comment for a_char22'
 
)
  comment 'I AM Table Comment22';
~~~


#### sqlite3
* ##### not supported
* warning
~~~bash 
⚡ python manage.py makemigrations     
System check identified some issues:

WARNINGS:
hello1.AModel.a_char: (field.W163) sqlite does not support a db_comment on columns
hello1.AModel.a_int: (field.W163) sqlite does not support a db_comment on columns
hello1.AModel: (models.W045) sqlite does not support db_table_comment.
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, hello1, sessions
Running migrations:
  Applying admin.0004_alter_logentry_content_type... OK
  Applying auth.0013_alter_permission_content_type... OK
  Applying hello1.0003_alter_amodel_a_array_alter_amodel_a_char_and_more... OK
  ...

~~~

jchubber commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,24 @@ characters that aren't allowed in Python variable names -- notably, the
+``db_comment``
+--------------
+
+.. attribute:: Field.db_comment
```
The documentation says that new features should have the `versionadded` block ([documentation](https://docs.djangoproject.com/en/3.2/internals/contributing/writing-documentation/#documenting-new-features)) like this:
```
.. versionadded:: 4.0
```
(It looks like the current in-development version is 4.0: [link](https://code.djangoproject.com/wiki/Version4.0Roadmap))
But I'm not 100% clear on whether the author should be adding that block or if that's the job of the release manager (in this case, Mariusz Felisiak). I'll post that question into the django-developers discussion group.

jchubber replied:
Update: Carlton confirmed that adding the versionadded block is appreciated.

jchubber commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,24 @@ characters that aren't allowed in Python variable names -- notably, the
+
+.. attribute:: Field.db_comment
+
+db_comment support SQL COMMENT.
+
+It is similar to `help_text`_ in terms of displaying. it's useful for sharing
+documents with DBA as well as developers. If you want to share `help_text`_
+with DBA, use ``db_comment`` instead of `help_text`_. For example::
+
+    db_comment="this is denormalization Field, do not update this field alone."
```
Suggesting some slight wording changes and the addition of a warning:
```
A string describing the Model field which will be saved as a SQL 
COMMENT on the SQL COLUMN

Unlike `help_text`, db_comments are displayed in the database as
SQL COMMENTs, which makes them useful for documenting fields
for database administrators, developers, analysts, and others with
direct database access, especially if you need to provide different
explanations  for users or site admins (using `help_text`) vs. for DBAs
(using `db_comment`). For example::    

    db_comment="User karma points. This is field is denormalized,\
                          so should not be updated alone. This comment is\
                          managed by Django, and will be updated on every\
                          database migration."

db_comment has a character limit that differs for each database 
backend. PostgreSQL has a 1GB limit. Oracle has a _____ limit. 
MySQL has a 1024 character limit. 

.. warning::
    Changes to database column comments made by database 
    users outside of Django will be overwritten by Django migrations 
    if Django finds a value for db_comment on a model field. 
    Communicating with your DBAs and issuing reminder notices in 
    the comments (as in the example above) may help avoid conflicts.
```

jchubber commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,15 @@ Django quotes column and table names behind the scenes.
+--------------------
+
+.. attribute:: Options.db_table_comment
+
```
Possibly add `.. versionadded: 4.0` block

jchubber commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,15 @@ Django quotes column and table names behind the scenes.
+
+.. attribute:: Options.db_table_comment
+
+    The database table comment to use for the model::
```
Some additional suggestions for this part of the documentation:

```
A string describing the Model which will be saved as a SQL 
COMMENT on the SQL TABLE

db_table_comments are displayed in the database as SQL 
COMMENTs, which makes them useful for documenting database 
tables for database administrators, developers, analysts, and others 
with direct database access who may not be looking at your Django
code.
For example::    

    db_table_comment="Each entry in this table is gathered when\
                          the user registers. This comment is managed by\
                          Django, and will be updated on every migration."

db_table_comment has a character limit that differs for each database 
backend. PostgreSQL has a 1GB limit. Oracle has a _____ limit. 
MySQL has a 2048-character limit. 

.. warning::
    Changes to database table comments made by database 
    users outside of Django will be overwritten by Django migrations 
    if Django finds a value for db_table_comment on a model. 
    Communicating with your DBAs and issuing reminder notices in 
    the comments (as in the example above) may help avoid conflicts.

jchubber reviewed (changes requested):
This is great! Very exciting to see the feature almost ready. In addition to the specific comments on specific lines, this PR also needs:
1. Test coverage.
2. An entry added to the [version 4.0 release notes](https://github.com/django/django/blob/main/docs/releases/4.0.txt)

@KimSoungRyoul I would be happy to work with you on any / all of these changes. If you'd like me to put together a PR on your branch to incorporate the recommended changes below, and to include some unit tests and the release note update, please let me know and I'd be happy to put that together this week for you to review / accept / update this PR with.

KimSoungRyoul commented:
thank you for your feedback  😄 

I will  add testcase for this feature and reflect your feedback until this weekend

KimSoungRyoul commented:
Hi 
It tries to be helpful in writing the test case. In this case, the new SQL grammar is reflected in django, so it can not be written with `assertTableExists()` or `assertColumnExists()` and other test utils. 
In order to make test cases, it is necessary to add information about `db_comment` to `DatabaseWrapper.introspection.get_table_description()`.  
However, I am not sure whether it is the right way to change `get_table_description()` method to only write a test case.

jchubber commented:
@KimSoungRyoul I do not have enough familiarity with migration testing either. I'll write a request into the `Django Core Mentorship` or `Django developers` groups to request advice. I'll let you know if/when I hear back.

KimSoungRyoul commented:
thank you for your feedback ,  yes  I  already reflected every piece of feedback.
I'll think about testcase for this new feature without modifying `DatabaseWrapper.introspection.get_table_description()` 

and I will  do rollback 1 commit and force push which contained failed testcase

jchubber commented:
@KimSoungRyoul You posted to django-developers and I posted to Django Core Mentorship ([link to thread](https://groups.google.com/g/django-core-mentorship/c/3KHlBAfpD5I)). It looks like the feedback from both groups is consistent: extending `get_table_description()` is ok and it's also recommended to extend inspectdb as well. It's great to have confirmation from both discussion boards about that ✅ ✅ 👍
If you need any help or want to discuss further, please don't hesitate to let me know.

KimSoungRyoul commented:
I haven't done much make testcase yet.
p.s
I would be grateful if you could advise on a test case scenario.

knyghty commented on django/db/backends/base/features.py:
```diff
@@ -24,6 +24,7 @@ class BaseDatabaseFeatures:
     supports_partially_nullable_unique_constraints = True
     # Does the backend support initially deferrable unique constraints?
     supports_deferrable_unique_constraints = False
+    supports_db_comments = False
```
I think the `_db` is redundant here - it's quite clear it refers to the database since we're in database features.

knyghty commented on django/db/backends/oracle/schema.py:
```diff
@@ -120,6 +126,7 @@ def _alter_field_type_workaround(self, model, old_field, new_field):
             self.quote_name(new_temp_field.column),
             new_value,
         ))
+
```
This extra newline seems unrelated and should be removed.

knyghty commented on tests/requirements/postgres.txt:
```diff
@@ -1 +1 @@
-psycopg2>=2.5.4
+psycopg2-binary>=2.5.4
```
I think you left this in.

knyghty reviewed (commented):
This looks like roughly the right approach to me but I don't know the schema generation code well enough to make much comment otherwise.

For the tests, I'm not sure the test you have is the best way to do it. I think maybe having a look at `tests/schema/tests.py` and adding something there would be more appropriate, but I'm not sure. Probably it would be good to test the system checks work as well, looking at `tests/invalid_models_tests/test_ordinary_fields.py` may help. Since you've added introspection code you'll also need to test that, maybe in `tests/inspectdb/tests.py`.

knyghty commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,37 @@ Django quotes column and table names behind the scenes.
+db_table_comment has a character limit that differs for each database
+backend. PostgreSQL has a 1GB limit. MySQL has a 2048-character limit.
+
+.. warning::
+    Changes to database table comments made by database
+    users outside of Django will be overwritten by Django migrations
+    if Django finds a value for db_table_comment on a model.
+    Communicating with your DBAs and issuing reminder notices in
+    the comments (as in the example above) may help avoid conflicts.
```
I don't think this warning is necessary - this is how Django works for everything else, I don't know if we need to call it out for comments specifically.

KimSoungRyoul commented:
@knyghty thank you for your feedback. 
It seems more appropriate to remove `.. warning::`. 
and `supports_db_comments ` -> `supports_comments `

@jchubber 
I add more testcase to cover up new feature(`db comment`)
* add_db_comment
* alter_db_comment
* delete_db_comment
* add_db_comment_with_default
* alter_db_comment_with_default

---


Is there anything else to think about besides the feedback you suggested?

knyghty commented:
What happens if you add `db_comment` to a field that isn't directly represented as a column, e.g. `ManyToManyField`? Do we need a system check or something for this? You may need to add something here: https://github.com/django/django/blob/main/django/db/models/fields/related.py#L1230-L1261 ideally with a test.

I'm not sure if any other fields will have this issue.

KimSoungRyoul commented:
I didn't consider about that. 
Just like you said,  we need a system check  
like a ...
~~~Python
if self.db_comment:
            warnings.append(
                checks.Warning(
                    'db_comment has no effect on ManyToManyField.',
                    obj=self,
                    id='fields.W3xx',
                )
            )

~~~

I will add this and consider what happens when `ManyToManyField` or `GenericForeignKey`  etc... and `db_comment` are used together.  😄 

Thanks

jchubber commented:
(👍 awesome work @KimSoungRyoul and great points @knyghty !)

**ManyToManyFields:** This was a good catch by @knyghty. Django creates an intermediary join table in the db for every ManyToMany field. In theory, we could support db_comments on `ManyToManyFields` by putting those comments as **_table-level_** comments onto the intermediary join table that Django created. HOWEVER, we'd encounter a problem when the user specified a ManyToManyField with a `through` model, if they tried to put a db_comment on both the Field and also on the through model definition. I'm not 100% sure how to handle this. Either (1) Django supports db_comments on ManyToManyFields and adds them to the intermediary join table as a table-level comment _when there is NO `through` table_, but ignores the field-level comment _when there IS a `through` table_, or (2) Django never supports db_comments on ManyToManyFields at all (basically "If you want a db comment on your M2M fields, you HAVE to use a through table and put the db_comment there"). @KimSoungRyoul What do you think?

jchubber commented:
**Test Cases:** 
The test cases @KimSoungRyoul has now look good. I am trying to think of strange edge cases situations where covering with unit tests could be helpful. I'm not coming up with many. 

- ManyToManyField comments: As mentioned above, this special case might require separate unit tests.
- Table-level comments on `through` models: Is there any difference for `through` models? I suppose not... Maybe unit tests on the `through` models is not needed?
- What's the expected result if there is a comment in the db on a column and the Django user does NOT specify any comment  on the field? Django should make no change to the pre-existing comment, right?

jchubber commented:
I did an additional review of the Django model docs and found three other situations where db_comment support might get complicated:

### 1) Unmanaged models:
([link to docs](https://docs.djangoproject.com/en/dev/ref/models/options/#managed)). For any model that has...
```python
class Meta:
        managed = False
```
...Django will not perform create, modify, or delete operations on the model. That means that db_comments cannot be supported for unmanaged models. I recommend that the docs be updated to state that db_comments are ignored when managed = False. 

### 2) Proxy models:
([link to docs](https://docs.djangoproject.com/en/dev/topics/db/models/#proxy-models)). A `proxy` model doesn't exist in the db in the same way as a regular model. So if a model has...

```python
class Meta:
        proxy = True
```
...then I think perhaps db_comments should not be supported at the table _or_ column-levels. That should be documented in the docs, should trigger a warning (during migration?), and should have a unit test to ensure that if a Django user specifies a db_comment on a proxy model that Django doesn't error.

### 3) Abstract Base Classes:
([link to docs](https://docs.djangoproject.com/en/3.2/topics/db/models/#abstract-base-classes)) Django doesn't create tables for Abstract Base Classes, so a table-level db_comment on an Abstract Base Class **_does NOT_** make sense. But a field-level (or "column-level") db_comment on an Abstract Base Class' fields _**DOES**_ make sense because those fields will eventually be created for any of the models that inherit from the Abstract Base Class.

So for example, with...

```python
from django.db import models

class CommonInfo(models.Model):
    name = models.CharField(max_length=100)
    age = models.PositiveIntegerField(db_comment="User's age when they registered.")

    class Meta:
        abstract = True

class Student(CommonInfo):
    home_group = models.CharField(max_length=5)
```
...then a unit test should verify that in the db the `student` table has a column `age` with the comment "`User's age when they registered.`".

jchubber commented:
@KimSoungRyoul I was busy the past 2 weeks, but I have more control over my time for the coming 2 weeks. Is there anything here that I can be helpful with? Writing docs? Working on code for some piece of the items mentioned above 👆?

KimSoungRyoul commented:
> (👍 awesome work @KimSoungRyoul and great points @knyghty !)
> 
> **ManyToManyFields:** This was a good catch by @knyghty. Django creates an intermediary join table in the db for every ManyToMany field. In theory, we could support db_comments on `ManyToManyFields` by putting those comments as **_table-level_** comments onto the intermediary join table that Django created. HOWEVER, we'd encounter a problem when the user specified a ManyToManyField with a `through` model, if they tried to put a db_comment on both the Field and also on the through model definition. I'm not 100% sure how to handle this. Either (1) Django supports db_comments on ManyToManyFields and adds them to the intermediary join table as a table-level comment _when there is NO `through` table_, but ignores the field-level comment _when there IS a `through` table_, or (2) Django never supports db_comments on ManyToManyFields at all (basically "If you want a db comment on your M2M fields, you HAVE to use a through table and put the db_comment there"). @KimSoungRyoul What do you think?


I think `(2) Django never supports db_comments on ManyToManyFields at all` is a good choice. 
`(1)`  is a possibility that it will become too complicated and I agree with the problem you mentioned.

Logically, it is correct that ManyToManyField does not support `db_comment` .  
**ManyToManyField is not a column. It's a relationship mapping between models.**

KimSoungRyoul commented:
Why fail to connect only py3.9-mysql?

Is it just a temporary issue?

~~~Python
  File "/home/jenkins/workspace/pull-requests-bionic/database/mysql_gis/label/bionic-pr/python/python3.9/tests/.env/lib/python3.9/site-packages/MySQLdb/__init__.py", line 130, in Connect
    return Connection(*args, **kwargs)
  File "/home/jenkins/workspace/pull-requests-bionic/database/mysql_gis/label/bionic-pr/python/python3.9/tests/.env/lib/python3.9/site-packages/MySQLdb/connections.py", line 185, in __init__
    super().__init__(*args, **kwargs2)
django.db.utils.OperationalError: (2002, "Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)")
Build step 'Execute shell' marked build as failure

~~~

KimSoungRyoul commented:
> @KimSoungRyoul I was busy the past 2 weeks, but I have more control over my time for the coming 2 weeks. Is there anything here that I can be helpful with? Writing docs? Working on code for some piece of the items mentioned above 👆?

thanks for your feedback
I thought most of your suggestions were appropriate, so I reflected most of them. Are there more areas that need improvement?  If not, may I ask who can advice to get approval?

knyghty commented on django/db/backends/base/schema.py:
```diff
@@ -732,6 +738,15 @@ def _alter_field(self, model, old_field, new_field, old_type, new_type,
+
+        if old_field.db_comment != new_field.db_comment:
+            fragment, other_actions = self._alter_column_comment_sql(model, new_field, new_type, new_field.db_comment)
+            # postgres, oracle can't execute same time 'ALTER COLUMN ...' and 'COMMENT ON ...'
```
I'm not exactly sure why this comment is here, I don't think it adds to readability. If we keep it, we can improve the English a bit:

```suggestion
            # Postgres and Oracle can't execute 'ALTER COLUMN ...' and 'COMMENT ON ...' at the same time.
```

knyghty commented on django/db/backends/base/schema.py:
```diff
@@ -929,6 +945,10 @@ def _alter_column_type_sql(self, model, old_field, new_field, new_type):
         )
 
+    def _alter_column_comment_sql(self, model, new_field, new_type, new_db_comment):
+        logger.debug("%s does not support db_comment sql, db_table_comment was ignored", self.connection.vendor)
```
```suggestion
        logger.debug("%s does not support db_comment, db_table_comment was ignored", self.connection.vendor)
```

knyghty commented on django/db/models/fields/__init__.py:
```diff
@@ -316,6 +318,18 @@ def _check_db_index(self):
+        if self.db_comment and not connection.features.supports_comments:
+            return [
+                checks.Warning(
+                    '{vendor} does not support a db_comment on columns'.format(vendor=connection.display_name),
```
```suggestion
                    '{vendor} does not support db_comment on columns'.format(vendor=connection.display_name),
```

knyghty commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,41 @@ characters that aren't allowed in Python variable names -- notably, the
 scenes.
 
+
+``db_comment``
```
I think this whole section should be moved underneath the others, I think the others are more important and things should generally be added at the bottom.

jchubber replied:
The list is alphabetized. This should remain in its current location so that the list remains alphabetized.

knyghty commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,41 @@ characters that aren't allowed in Python variable names -- notably, the
+.. versionadded:: 4.0
+.. attribute:: Field.db_comment
+
+db_comment support SQL COMMENT.
```
I think we can remove this whole line.

knyghty commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,41 @@ characters that aren't allowed in Python variable names -- notably, the
+COMMENT on the SQL COLUMN
+
+Unlike `help_text`, db_comments are displayed in the database as
+SQL COMMENTs, which makes them useful for documenting fields
```
```suggestion
SQL comments, which is useful for documenting fields
```

knyghty commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,41 @@ characters that aren't allowed in Python variable names -- notably, the
+db_comment has a character limit that differs for each database
+backend. PostgreSQL has a 1GB limit. MySQL has a 1024 character limit.
+
+.. warning::
+    Changes to database column comments made by database
+    users outside of Django will be overwritten by Django migrations
+    if Django finds a value for db_comment on a model field.
+    Communicating with your DBAs and issuing reminder notices in
+    the comments (as in the example above) may help avoid conflicts.
```
I think I mentioned this before, but as it applies to more or less everything in Django, I think we don't need it.

jchubber replied:
I agree. This can be removed.

knyghty commented on docs/topics/db/models.txt:
```diff
@@ -1356,6 +1356,20 @@ The general rules are:
    This sets things up so that the proxy model is an exact copy of the
    storage structure of the original model when data is saved.
 
+
+Proxy model and Unmanaged models ignore DB comment options
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If Model has ``Meta.proxy=True``. or ``Meta.managed=False``,
+``Meta.db_table_comment=table comment...`` and ``Field.db_comment=comment...``
+has no effect.
+
+therefore it is recommended not to use ``Meta.proxy=True`` with
+``Meta.db_table_comment=table comment...`` and ``Field.db_comment=comment...``
+likewise it is recommended not to use ``Meta.managed=False`` with
+``Meta.db_table_comment=table comment...`` and ``Field.db_comment=comment...``.
+
+
```
I don't think we need this, as it applies equally to any database changes in proxy models.

jchubber replied:
While I agree that this `Proxy model and Unmanaged models ignore DB comment options` section probably shouldn't sit in this location in the docs, I do think it's useful to have this behavior documented in the section of the docs where Meta.db_table_comment and Field.db_comment are introduced. Because although experts will be familiar with the behavior of proxymodels, the interaction is not obvious for newer Django users.

KimSoungRyoul replied:
I agree that the location of this is not appropriate, and I will delete it because it seems to need rewrite this comment because  it is awkward sentence

knyghty commented on docs/releases/4.0.txt:
```diff
@@ -58,6 +58,45 @@ For example::
 :attr:`Meta.constraints <django.db.models.Options.constraints>` option.
 
+
+Supoort SQL COMMENT
```
I'm not sure this is enough to count as a major feature IMO (other opinions may vary) - if it does, I think we don't need the note about ManyToManyField here, just the first example is enough I think.

jchubber replied:
Not sure what makes a feature "major" vs. "minor", but I'd like to share what I think makes this feature _at least notable_:

- This feature is among the first Django features targeted primarily at database administrators, data analysts, data scientists, and others who are looking to consume data that is managed by Django.
- The idea for this feature is 11 years old (if you go back to ticket #13867 which is the predecessor for ticket #18468 that this patch will close).

I'm not a committer so I'll defer to committers on what the "bar" is for "major feature" status...

KimSoungRyoul replied:
I agree that it's not a major feature but I think at least it needs to be mentioned because it is a new feature.
I am not familiar with django's policy, so if it seems inappropriate for releases note to comment here, please leave one more comment.

knyghty reviewed (commented):
I left a few comments, mostly around the docs / grammar. They're all quite subjective.

I think you're right about ManyToManyField, we don't need to add a maintenance burden here. If a user wants to add a comment to the table, this is already supported: they can use a through model or a SQL migration.

knyghty commented:
@charettes, would you like to take a look?

jchubber commented:
> I thought most of your suggestions were appropriate, so I reflected most of them. Are there more areas that need improvement? If not, may I ask who can advice to get approval?

Because this patch now meets every requirement on the development checklist, I've updated the status in the Django ticket tracker so that it shows up in the `Patches Needing Review` section of the [Django Development Dashboard](https://dashboard.djangoproject.com/).

The next step is for anyone except the patch author(s) to review the patch using [the patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) and either mark the ticket as `Ready for checkin` on the [Django Development Ticket Tracker Entry for Ticket #18468](https://code.djangoproject.com/ticket/18468) if everything looks good, or leave comments for improvement and mark the ticket as `Patch needs improvement`. I'd also request that comments for improvement be left here in Github as well.

KimSoungRyoul commented:
> I left a few comments, mostly around the docs / grammar. They're all quite subjective.

Grammar is my weakness, so I'll reflect all the feedback involved .
thanks!

KimSoungRyoul commented:
Hi may I know what more I should do? such as requesting reviews from others or reflecting feedback

jchubber commented:
@KimSoungRyoul I'm reviewing the using [this patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist). Here's the current status of that checklist:

- [x] Documentation: Does the documentation build without any errors (make html, or make.bat html on Windows, from the docs directory)? `Yes, it builds with no errors ✅` 
- [ ] Documentation: Does the documentation follow the writing style guidelines in Writing documentation? `There are two warnings. See below in the list of remaining changes. 🟡`
- [x] Documentation: Are there any spelling errors? `None. Looks good ✅`
- [x] Bugs: Is there a proper regression test (the test should fail before the fix is applied)? `Yes, the tests will fail before the fix is applied ✅`
- [x] Bugs: If it’s a bug that qualifies for a backport to the stable version of Django, is there a release note in docs/releases/A.B.C.txt? Bug fixes that will be applied only to the main branch don’t need a release note. `Not applicable. No backport needed.  ✅`
- [x] New Features: Are there tests to “exercise” all of the new code? `Yes. Well exercised. ✅`
- [x] New Features: Is there a release note in docs/releases/A.B.txt? `Yes there is ✅`
- [x] New Features: Is there documentation for the feature and is it annotated appropriately with .. versionadded:: A.B or .. versionchanged:: A.B? `Yes.  ✅`
- [x] All code changes: Does the coding style conform to our guidelines? Are there any flake8 errors? You can install the pre-commit hooks to automatically catch these errors. `I ran flake8 on all files included in this patch and found no errors. ✅`
- [x] If the change is backwards incompatible in any way, is there a note in the release notes (docs/releases/A.B.txt)? `Not applicable. This change does not lead to any backwards incompatibility. ✅`
- [x] Is Django’s test suite passing? `Not yet... I got one failure, which I'm investigating. I'm not sure if it's related to this patch or not. I'll report back. ⏱`
- [ ] All tickets: Is the pull request a single squashed commit with a message that follows our commit message format? `It's not a single squashed commit, but it does have a message that follows our commit message format. I'm not sure how important it is for the patch to be a single squashed commit, but if it's important, then I think KimSoungRyoul is the person who should squash the commit (so you retain authorship). If you are unfamiliar with squashing a commit (without having to create a new PR) then I'll go ahead and mark this as ready for merging and we'll see if the committers are comfortable with not having a single squashed commit. 🟡`
- [ ] All tickets: Are you the patch author and a new contributor? Please add yourself to the AUTHORS file and submit a Contributor License Agreement. `No. I do not see that @KimSoungRyoul added themselves to the AUTHORS file and I am unsure if they submitted a Contributor License Agreement. @KimSoungRyoul should do both. 🟡` (@KimSoungRyoul here's a link for the contributor license agreement instructions https://www.djangoproject.com/foundation/cla/)

## Changes that remain to be made:
- [This](https://github.com/django/django/pull/14463/files#diff-fc688171203d4ce1e42c46f25bddfee5f494b0b1b81258907ff0531c81f2a4aeR332) should be ``` ``help_text`` ``` (not ``` `help_text` ```)
- [This](https://github.com/django/django/pull/14463/files#diff-fc688171203d4ce1e42c46f25bddfee5f494b0b1b81258907ff0531c81f2a4aeR337) should be ``` ``db_comment`` ``` (not ``` `db_comment ` ```)
- The warning on the db_comment docs [here](https://github.com/django/django/pull/14463/files#diff-fc688171203d4ce1e42c46f25bddfee5f494b0b1b81258907ff0531c81f2a4aeR347-R352) should be removed
- Adding yourself to the `AUTHORS` file here: https://github.com/django/django/blob/main/AUTHORS .

Once these changes are made, I'll mark this "Ready for checkin" and then a committer will have the chance to review and either accept it or send it back for changes. (I'm not a committer)

jchubber commented on tests/invalid_models_tests/test_relative_fields.py:
```diff
@@ -115,6 +116,11 @@ class ThroughModel(models.Model):
                 obj=field,
                 id='fields.W340',
             ),
+            DjangoWarning(
+                'db_comment has no effect on ManyToManyField.',
+                obj=field,
+                id='fields.W345',
+            ),
```
This test is failing when I run the test suite locally. Here's the error message:
```
FAIL: test_many_to_many_with_useless_options (invalid_models_tests.test_relative_fields.RelativeFieldTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jaredchung/Documents/workspace/django-dbcomment/django/tests/invalid_models_tests/test_relative_fields.py", line 113, in test_many_to_many_with_useless_options
    self.assertEqual(ModelM2M.check(), [
AssertionError: Lists differ: [<War[16 chars]msg='SQLite does not support db_comment on col[759 chars]45'>] != [<War[16 chars]msg='null has no effect on ManyToManyField.', [597 chars]43'>]

First differing element 0:
<Warn[15 chars]msg='SQLite does not support db_comment on col[89 chars]163'>
<Warn[15 chars]msg='null has no effect on ManyToManyField.', [83 chars]340'>

First list contains 1 additional elements.
First extra element 4:
<Warning: level=30, msg='db_comment has no effect on ManyToManyField.', hint=None, obj=<django.db.models.fields.related.ManyToManyField: m2m>, id='fields.W345'>

Diff is 1758 characters long. Set self.maxDiff to None to see it.
```

jchubber reviewed (changes requested):
One test failure: `test_many_to_many_with_useless_options`

KimSoungRyoul commented:
all done

* fix documentation follow the writing style ( and remove)
* fix failing testcase 
* adding to the AUTHORS (and submit a Contributor License Agreement.)

Thanks again for your guide. @jchubber  :bowing_man:

jchubber commented:
# Updated checklist status:

- [x] Documentation: Does the documentation build without any errors (make html, or make.bat html on Windows, from the docs directory)? `Yes, it builds with no errors ✅` 
- [x] Documentation: Does the documentation follow the writing style guidelines in Writing documentation?  `Looks good. ✅` 
- [x] Documentation: Are there any spelling errors? `None. Looks good ✅`
- [x] Bugs: Is there a proper regression test (the test should fail before the fix is applied)? `Yes, the tests will fail before the fix is applied ✅`
- [x] Bugs: If it’s a bug that qualifies for a backport to the stable version of Django, is there a release note in docs/releases/A.B.C.txt? Bug fixes that will be applied only to the main branch don’t need a release note. `Not applicable. No backport needed.  ✅`
- [x] New Features: Are there tests to “exercise” all of the new code? `Yes. Well exercised. ✅`
- [x] New Features: Is there a release note in docs/releases/A.B.txt? `Yes there is ✅`
- [x] New Features: Is there documentation for the feature and is it annotated appropriately with .. versionadded:: A.B or .. versionchanged:: A.B? `Yes.  ✅`
- [x] All code changes: Does the coding style conform to our guidelines? Are there any flake8 errors? You can install the pre-commit hooks to automatically catch these errors. `I ran flake8 on all files included in this patch and found no errors. ✅`
- [x] If the change is backwards incompatible in any way, is there a note in the release notes (docs/releases/A.B.txt)? `Not applicable. This change does not lead to any backwards incompatibility. ✅`
- [x] Is Django’s test suite passing? `Passing tests. ✅`
- [x] All tickets: Is the pull request a single squashed commit with a message that follows our commit message format? `Yes, it's a single squashed commit. ✅`
- [ ] All tickets: Are you the patch author and a new contributor? Please add yourself to the AUTHORS file and submit a Contributor License Agreement. `Yes, @KimSoungRyoul added themselves to the AUTHORS file ✅`

# 👉  Patch Review Checklist is Complete ✅  

I have marked this patch `Ready for checkin` in the Django ticket tracker. A committer will have the chance to review and either accept it or send it back for changes. (I'm not a committer)

felixxm commented on django/db/backends/oracle/introspection.py:
```diff
@@ -121,15 +121,18 @@ def get_table_description(self, cursor, table_name):
-        """, [table_name])
+            LEFT OUTER JOIN
+                user_col_comments ON user_col_comments.column_name = user_tab_cols.column_name
+            WHERE user_tab_cols.table_name = UPPER(%s) and user_col_comments.table_name = UPPER(%s)
```
You can `JOIN` with `user_col_comments` on multiple columns there is no need to use `UPPER()` again and pass `table_name` multiple times.

KimSoungRyoul replied:
I fixed this query to below

~~~SQL
...
                END as is_json,
                user_col_comments.comments as col_comment
            FROM user_tab_cols
            LEFT OUTER JOIN
                user_tables ON user_tables.table_name = user_tab_cols.table_name
            LEFT OUTER JOIN
                user_col_comments
                ON user_col_comments.column_name = user_tab_cols.column_name
                    and user_col_comments.table_name = %s
            WHERE user_tab_cols.table_name = %s

~~~

felixxm commented on django/db/backends/oracle/introspection.py:
```diff
@@ -121,15 +121,18 @@ def get_table_description(self, cursor, table_name):
         field_map = {
             column: (internal_size, default if default != 'NULL' else None, collation, is_autofield, is_json)
-            for column, default, collation, internal_size, is_autofield, is_json in cursor.fetchall()
+            for column, default, collation, internal_size, is_autofield, is_json, column_comment, in cursor.fetchall()
```
`column_comment` -> `comment`

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -21,6 +21,12 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     sql_create_column_inline_fk = 'CONSTRAINT %(name)s REFERENCES %(to_table)s(%(to_column)s)%(deferrable)s'
     sql_delete_table = "DROP TABLE %(table)s CASCADE CONSTRAINTS"
     sql_create_index = "CREATE INDEX %(name)s ON %(table)s (%(columns)s)%(extra)s"
+    sql_create_table_with_comment = """
+        CREATE TABLE %(table)s (%(definition)s);
+        COMMENT ON TABLE %(table)s IS '%(table_comment)s'
+    """
```
I'd rather execute additional statement based on `sql_create_table_comment` instead of adding another statement.

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -34,6 +40,30 @@ def quote_value(self, value):
             return str(value)
 
+    def create_model(self, model):
+        super().create_model(model)
+        if model._meta.db_table_comment:
+            self.execute(self.sql_create_table_comment % {
+                'table': model._meta.db_table,
+                'table_comment': model._meta.db_table_comment.replace('\'', '').replace('\n', ' ')
+            }, [])
```
This won't work on Oracle because you're trying to create the same table twice.

KimSoungRyoul replied:
I guess you may have misunderstood. sql was executed `self.sql_create_table_comment` not `sql_create_table_with_comment`

KimSoungRyoul replied:
these code was refacted 

~~~Python
 if model._meta.db_table_comment:
       self.execute(
            self.sql_alter_table_comment
            % {
                "table": model._meta.db_table,
                "comment": self.quote_comment_value(model._meta.db_table_comment),
            },
            [],
       )
~~~

felixxm commented on django/db/backends/sqlite3/introspection.py:
```diff
@@ -9,7 +9,7 @@
 from django.utils.regex_helper import _lazy_re_compile
 
-FieldInfo = namedtuple('FieldInfo', BaseFieldInfo._fields + ('pk', 'has_json_constraint'))
+FieldInfo = namedtuple('FieldInfo', BaseFieldInfo._fields + ('pk', 'has_json_constraint', 'db_comment'))
```
Why you added `db_comment` if it's not supported? :thinking:

KimSoungRyoul replied:
If it doesn't exists, `sqlite_connection.introspection.get_table_description(cursor, table_name)`  raise AttributeError() or "FieldInfo () args expected 12 but 13..."

felixxm reviewed (commented):
@KimSoungRyoul Thanks for this patch :+1: I left initial comments (only to schema and introspection) but it should be enough for now. There are also dozens of failures on Oracle.

What about introspection of table comments?

KimSoungRyoul commented:
@felixxm 

Hi I reflected your comments.  9d769edbbd6e008e4553a18fda4474e5d54fbcf6
I will squash all my commits after approve. thank you for your feedback

#### summary
* reusable `COMMENT` SQL 
    * define `def _comment_sql(self, comment):`
    
    *  remove comment related sql or method like a `sql_create_table_with_comment`
       and left comment related sql only `sql_alter_table_comment` and `sql_alter_column_comment`
    
* `column_comment` -> `comment`

*   Each database handles '', None differently. so I define `def quote_comment_value(self, value):`

jacobtylerwalls commented:
Hi @KimSoungRyoul, if this is ready for re-review, please uncheck "Needs improvement" on the Trac ticket, which is how we track waiting-for-author. While here, I noticed the release notes and version annotations need to be moved to 4.1, which is under dev now. Thanks!

claudep commented:
Hey @KimSoungRyoul, do you plan to continue improving this pull request as per comments?

sergun commented:
Hi @KimSoungRyoul KimSoungRyoul

Thanks a lot for your great job!

I have two questions:
1. Is it safe in your opinion to use your very useful implementation ia a patch (so download https://github.com/django/django/pull/14463.patch and do git apply for my local django installation) for django 4.0.x or 4.1.x?
2. Can you please advise some elegant way to use your implementation but to use verbose_name or help_text as a source for database comments for fields. I try to avoid specifying both things: e.g. help_text + db_comment

KimSoungRyoul commented on django/db/backends/mysql/introspection.py:
```diff
@@ -9,11 +9,12 @@
 
-FieldInfo = namedtuple('FieldInfo', BaseFieldInfo._fields + ('extra', 'is_unsigned', 'has_json_constraint'))
+FieldInfo = namedtuple('FieldInfo',
+                       BaseFieldInfo._fields + ('extra', 'is_unsigned', 'has_json_constraint', 'column_comment'))
```
now code format is depend on black :smiling_face_with_tear:

KimSoungRyoul commented on django/db/backends/mysql/schema.py:
```diff
@@ -49,6 +52,16 @@ def sql_rename_column(self):
+
+        if field.db_comment and not field.many_to_many:
+            sql += " COMMENT '%(db_comment)s' " % {
+                'db_comment': field.db_comment.replace('\'', '"').replace('\n', ' ')
```
it was changed like below

~~~python

 if field.db_comment and not field.many_to_many:
     sql += self._comment_sql(field.db_comment)
....

def _comment_sql(self, comment):
     return " COMMENT " + self.quote_comment_value(comment)
~~~

KimSoungRyoul commented on django/db/backends/base/schema.py:
```diff
@@ -456,6 +458,10 @@ def alter_db_table(self, model, old_db_table, new_db_table):
                 sql.rename_table_references(old_db_table, new_db_table)
 
+    def alter_db_table_comment(self, model, old_db_table_comment, new_db_table_comment):
+        logger.debug("%s does not support db table comment sql, db_table_comment was ignored", self.connection.vendor)
```
This feedback was reflected

KimSoungRyoul commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,32 @@ characters that aren't allowed in Python variable names -- notably, the
+.. versionadded:: 4.0
+.. attribute:: Field.db_comment
+
+A string describing the Model field which will be saved as a SQL
```
According to the [Oracle docs](http://docs.oracle.com/database/121/CNCPT/sqllangu.htm#CNCPT116) it is sequel, so “a SQL …” is the correct way

refer: https://oracle-base.com/blog/2015/01/02/a-sql-or-an-sql/

felixxm replied:
Guidelines from our docs are more important, see [Commonly used terms](https://docs.djangoproject.com/en/4.1/internals/contributing/writing-documentation/#commonly-used-terms).

KimSoungRyoul commented on docs/ref/models/fields.txt:
```diff
@@ -319,6 +319,32 @@ characters that aren't allowed in Python variable names -- notably, the
+SQL comments, which is useful for documenting fields
+for database administrators, developers, analysts, and others with
+direct database access, especially if you need to provide different
+explanations  for users or site admins (using `help_text`) vs. for DBAs
```
Thank you for your feedback on my terrible writing skill 😄

KimSoungRyoul commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,35 @@ Django quotes column and table names behind the scenes.
+.. versionadded:: 4.0
+.. attribute:: Options.db_table_comment
+
+A string describing the Model which will be saved as a SQL
```
https://github.com/django/django/pull/14463#discussion_r996432674

KimSoungRyoul commented on django/db/backends/base/schema.py:
```diff
@@ -929,6 +945,10 @@ def _alter_column_type_sql(self, model, old_field, new_field, new_type):
 
+    def _alter_column_comment_sql(self, model, new_field, new_type, new_db_comment):
+        logger.debug("%s does not support db_comment, db_table_comment was ignored", self.connection.vendor)
+        return None, []
```
this log was deleted 
thanks

KimSoungRyoul commented:
I almost forget this pullrequest 

thank you for reminding me @sergun 

---

https://github.com/django/django/pull/14463#issuecomment-941152900
It's a checklist for the above comment

- [x] (resolve comment which was reflected)
- [x] (Oracle: @felixxm found dozens of failures on Oracle  <-- fixed it)
- [x] ( update release notes 4.2)
- [x] (accept feedback)
- [x] (Please squash to 1 commit and update the PR with those changes)

GabeMillikan commented on AUTHORS:
```diff
@@ -551,6 +551,7 @@ answer newbie questions, and generally made Django that much better:
     Kieran Holland <http://www.kieranholland.com>
     kilian <kilian.cavalotti@lip6.fr>
     Kim Joon Hwan 김준환 <xncbf12@gmail.com>
+    Kim Soung Ryoul 김성렬 <kimsoungryoul.gmail.com>
```
```suggestion
    Kim Soung Ryoul 김성렬 <kimsoungryoul@gmail.com>
```
Thanks for the work you did here, looking forward to seeing it released🙂

felixxm commented:
buildbot, test on oracle.

smithdc1 commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,34 @@ Django quotes column and table names behind the scenes.
+``db_table_comment``
+--------------------
+
+.. versionadded:: 4.0
```
```suggestion
.. versionadded:: 4.2
```

smithdc1 commented on docs/ref/models/fields.txt:
```diff
@@ -325,6 +325,31 @@ characters that aren't allowed in Python variable names -- notably, the
+                          database migration."
+
+``db_comment`` has a character limit that differs for each database
+backend. PostgreSQL has a 1GB limit. MySQL has a 1024 character limit.
```
I wonder if this part. Ould be generalised?

As written I'm asking what about oracle / sqlite? Do these limits change between versions too? That could be all too much for the docs.

smithdc1 commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,34 @@ Django quotes column and table names behind the scenes.
+.. versionadded:: 4.0
+.. attribute:: Options.db_table_comment
+
+A string describing the Model which will be saved as a SQL
```
```suggestion
A string describing the Model which will be saved as an SQL
```

smithdc1 commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,34 @@ Django quotes column and table names behind the scenes.
+COMMENTs, which makes them useful for documenting database
+tables for database administrators, developers, analysts, and others
+with direct database access who may not be looking at your Django
+code.
+For example::
```
```suggestion
code. For example::
```

smithdc1 commented on docs/ref/models/fields.txt:
```diff
@@ -325,6 +325,31 @@ characters that aren't allowed in Python variable names -- notably, the
+
+Unlike ``help_text``, db_comments are displayed in the database as
+SQL comments, which is useful for documenting fields
+for database administrators, developers, analysts, and others with
```
```suggestion
for individuals with
```

smithdc1 commented on docs/ref/schema-editor.txt:
```diff
@@ -127,6 +127,17 @@ value.
+this will add or update comment from the models's table until they match the
+new value.
+
+
+
```
```suggestion
```

smithdc1 commented on docs/releases/4.2.txt:
```diff
@@ -376,6 +413,7 @@ but it should not be used for new migrations. Use
 :class:`~django.db.migrations.operations.AddIndex` and
 :class:`~django.db.migrations.operations.RemoveIndex` operations instead.
 
+
```
```suggestion
```

smithdc1 commented on docs/releases/4.2.txt:
```diff
@@ -229,6 +229,43 @@ Models
+-------------------
+
+* The new option :attr:`db_comment <django.db.models.Field.db_comment>` to
+  Field
```
```suggestion
  ``Field``.
```

smithdc1 commented on docs/releases/4.2.txt:
```diff
@@ -229,6 +229,43 @@ Models
+  Field
+
+* The new option :attr:`db_table_comment
+  <django.db.models.Options.db_table_comment>` to Model Meta
```
```suggestion
  <django.db.models.Options.db_table_comment>` to Model Meta.
```

smithdc1 commented on docs/releases/4.2.txt:
```diff
@@ -229,6 +229,43 @@ Models
+* The new option :attr:`db_table_comment
+  <django.db.models.Options.db_table_comment>` to Model Meta
+
+
```
```suggestion
```

smithdc1 reviewed (commented):
Just a few small thoughts on the docs.

KimSoungRyoul commented:
@smithdc1 
thank you for feedback 😄 


@felixxm It looks like a test run command. 
> buildbot, test on oracle.

may I ask how can I see the test results?

felixxm commented:
buildbot, test on oracle.

felixxm commented on django/db/backends/oracle/introspection.py:
```diff
@@ -131,13 +133,18 @@ def get_table_description(self, cursor, table_name):
+            LEFT OUTER JOIN
+                user_col_comments
+                ON user_col_comments.column_name = user_tab_cols.column_name
+                    and user_col_comments.table_name = UPPER(%s)
```
```suggestion
                    and user_col_comments.table_name = user_tab_cols.table_name
```

felixxm commented on django/db/migrations/operations/models.py:
```diff
@@ -529,6 +529,41 @@ def migration_name_fragment(self):
+        return self.database_forwards(app_label, schema_editor, from_state, to_state)
+
+    def describe(self):
+        return "Modify Table Comment to %s" % (self.db_table_comment,)
```
```suggestion
        return f"Alter {self.name_lower} comment"
```

felixxm commented on django/db/migrations/operations/models.py:
```diff
@@ -529,6 +529,41 @@ def migration_name_fragment(self):
         return "alter_%s_table" % self.name_lower
 
 
+class AlterModelTableComment(ModelOptionOperation):
```
We should also add `migration_name_fragment()` property, e.g.
```python
    @property
    def migration_name_fragment(self):
        return f"alter_{self.name_lower}_table_comment"
```

felixxm commented on django/db/models/fields/__init__.py:
```diff
@@ -196,6 +196,7 @@ def __init__(
         choices=None,
         help_text="",
         db_column=None,
+        db_comment=None,
```
We cannot add a new argument between the existing ones. Please move it the end.

felixxm commented on docs/ref/checks.txt:
```diff
@@ -400,6 +402,7 @@ Models
   expressions.
 * **models.W045**: Check constraint ``<constraint>`` contains ``RawSQL()``
   expression and won't be validated during the model ``full_clean()``.
+* **models.W045**: ``<database>`` does not support db_table_comment.
```
```suggestion
* **models.W045**: ``<database>`` does not support ``db_table_comment``.
```

felixxm commented on docs/ref/checks.txt:
```diff
@@ -315,6 +316,7 @@ Related fields
   the table name of ``<model>``/``<model>.<field name>``.
 * **fields.W345**: ``related_name`` has no effect on ``ManyToManyField`` with a
   symmetrical relationship, e.g. to "self".
+* **fields.W346**: ``ManyToManyField`` does not support ``db_comment``.
```
Documentation doesn't match the raised warning.

KimSoungRyoul replied:
Please check again if you missed 
![스크린샷 2022-11-04 오후 7 54 02](https://user-images.githubusercontent.com/24240623/199956118-e3c68411-5419-470c-8ed0-f788917381e4.png)

felixxm replied:
but the message is different

felixxm commented on docs/releases/4.2.txt:
```diff
@@ -229,6 +229,42 @@ Models
+                           the user registers. This comment is managed by\
+                           Django, and will be updated on every migration.'
+
+
+
+
```
Remove unnecessary blank lines.

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -28,6 +28,9 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     sql_delete_table = "DROP TABLE %(table)s CASCADE CONSTRAINTS"
     sql_create_index = "CREATE INDEX %(name)s ON %(table)s (%(columns)s)%(extra)s"
 
+    sql_alter_table_comment = "COMMENT ON TABLE %(table)s IS %(comment)s"
+    sql_alter_column_comment = "COMMENT ON COLUMN %(table)s.%(column)s IS %(comment)s"
```
We can remove these attributes as they are the same as in the base backend.

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +45,42 @@ def quote_value(self, value):
+                    self.sql_alter_column_comment
+                    % {
+                        "table": model._meta.db_table,
+                        "column": field.db_column or field.get_attname(),
```
Why not `field.column`?
```suggestion
                        "column": field.column,
```

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +45,42 @@ def quote_value(self, value):
+                self.sql_alter_column_comment
+                % {
+                    "table": model._meta.db_table,
+                    "column": field.db_column or field.get_attname(),
```
Why not `field.column`?
```suggestion
                    "column": field.column,
```

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -250,3 +289,16 @@ def _alter_column_collation_sql(
+    def _alter_column_comment_sql(self, model, new_field, new_type, new_db_comment):
+        return None, [
+            (
+                self.sql_alter_column_comment
+                % {
+                    "table": self.quote_name(model._meta.db_table),
+                    "column": self.quote_name(new_field.column),
+                    "comment": self.quote_comment_value(new_db_comment),
+                },
+                [],
```
This (or similar code) is generated in many places in all backends. We should add an internal hook to share this logic, e.g. `BaseDatabaseSchemaEditor._alter_column_comment_sql()`.

felixxm replied:
IMO, `_alter_column_comment_sql()` should return only `sql`, `params`. This would allow it to reused it other places, e.g. in `create_model()` or `add_field()` on Oracle:
```python
def _alter_column_comment_sql(self, model, new_field, new_type, new_db_comment, old_db_comment):
    if (
        not self.connection.features.supports_comments
        or new_field.many_to_many
        or old_db_comment == new_db_comment
    ):
        return "", []
    return (
        self.sql_alter_column_comment
        % {
            "table": self.quote_name(model._meta.db_table),
            "column": self.quote_name(new_field.column),
            "comment": self.quote_comment_value(new_db_comment),
        },
        [],
    )
```

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +45,42 @@ def quote_value(self, value):
+
+    def add_field(self, model, field):
+        super().add_field(model, field)
+        if field.db_comment and not field.many_to_many:
```
I'd move this check to the `BaseDatabaseSchemaEditor._alter_column_comment_sql()`.

felixxm reviewed (commented):
@KimSoungRyoul Thanks for updates :+1:

KimSoungRyoul commented:
@felixxm 
thank you for feedback 
I will reflect your feedback before tomorrow evening

KimSoungRyoul commented:
buildbot, test on oracle.

KimSoungRyoul commented:
buildbot, test on oracle.

KimSoungRyoul commented:
buildbot, test on oracle.

KimSoungRyoul commented:
Is it difficult to include 4.2-alpha release?

felixxm commented:
> Is it difficult to include 4.2-alpha release?

Django 4.2 feature freeze is on January 16, 2023. Don't worry this patch is on my list for Django 4.2.

ngnpope commented on django/db/backends/postgresql/introspection.py:
```diff
@@ -5,7 +5,14 @@
-FieldInfo = namedtuple("FieldInfo", BaseFieldInfo._fields + ("is_autofield",))
+FieldInfo = namedtuple(
+    "FieldInfo",
+    BaseFieldInfo._fields
+    + (
+        "is_autofield",
+        "comment",
+    ),
```
```suggestion
    BaseFieldInfo._fields + ("is_autofield", "comment"),
```

ngnpope commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +44,41 @@ def quote_value(self, value):
 
+    def create_model(self, model):
+        super().create_model(model)
+        if model._meta.db_table_comment:
+            self.execute(
+                self.sql_alter_table_comment
+                % {
+                    "table": model._meta.db_table,
+                    "comment": self.quote_comment_value(model._meta.db_table_comment),
+                },
+                [],
+            )
+        for field in model._meta.local_fields:
+            if field.db_comment:
+                self.execute(
+                    self.sql_alter_column_comment
+                    % {
+                        "table": model._meta.db_table,
+                        "column": field.column,
+                        "comment": self.quote_comment_value(field.db_comment),
+                    },
+                    [],
+                )
```
This looks the same for all backends that support this? Can we not just check the `supports_comments` flag in base class' `create_model()` method?

KimSoungRyoul replied:
mysql also supports db_comment, but there is only a difference in the sql grammar with postgresql & oracle
so it's hard to check flag `supports_comments`

ngnpope replied:
So I'm not sure it's worth having the complexity of allowing the comment to be defined inline on table creation or column addition. This is something that only MySQL/MariaDB supports, but they also support modifying the comment after creation anyway, so why not just do that consistently? We also don't want to support a hypothetical case where modification isn't possible.

If we were to keep this capability of "inline" comments I'd like to see an additional feature flag `supports_comments_inline` which can be use to toggle off the behaviour in `create_model()` and `add_field()` in the base backend. This allows us to remove everything custom from the Oracle and PostgreSQL backends other than the necessary `supports_comments` feature flag and the introspection bits.

Anyway, I'd prefer just ripping out the "inline" comments stuff.

felixxm replied:
> Anyway, I'd prefer just ripping out the "inline" comments stuff. What's your opinion @felixxm?

Agreed, it's much simpler when done the same way on all backends.
@KimSoungRyoul Does it work for you? see 670b718afdc310d33f9459d58fd6108641a3d9a2.

felixxm replied:
I had to restore the `supports_comments_inline` feature flag and partly also inline comments, to preserve column comments when altering column types (see also #16396).

ngnpope commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +44,41 @@ def quote_value(self, value):
+
+    def add_field(self, model, field):
+        super().add_field(model, field)
+        self.execute(
+            self.sql_alter_column_comment
+            % {
+                "table": model._meta.db_table,
+                "column": field.column,
+                "comment": self.quote_comment_value(field.db_comment),
+            },
+            [],
+        )
```
This looks the same for all backends that support this? Can we not just check the `supports_comments` flag in base class' `add_field()` method?

ngnpope reviewed (commented):
Some initial thoughts...

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +44,41 @@ def quote_value(self, value):
+                self.execute(
+                    self.sql_alter_column_comment
+                    % {
+                        "table": model._meta.db_table,
+                        "column": field.column,
```
Table and column names must be quoted.

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +44,41 @@ def quote_value(self, value):
+            self.execute(
+                self.sql_alter_table_comment
+                % {
+                    "table": model._meta.db_table,
```
Table and column names must be quoted.

felixxm commented on django/db/backends/oracle/schema.py:
```diff
@@ -42,6 +44,41 @@ def quote_value(self, value):
+        self.execute(
+            self.sql_alter_column_comment
+            % {
+                "table": model._meta.db_table,
+                "column": field.column,
```
Table and column names must be quoted.

felixxm commented on django/db/backends/postgresql/schema.py:
```diff
@@ -58,6 +58,44 @@ def quote_value(self, value):
+                self.execute(
+                    self.sql_alter_column_comment
+                    % {
+                        "table": model._meta.db_table,
+                        "column": field.column,
```
Table and column names must be quoted.

felixxm commented on django/db/backends/postgresql/schema.py:
```diff
@@ -58,6 +58,44 @@ def quote_value(self, value):
+        self.execute(
+            self.sql_alter_column_comment
+            % {
+                "table": model._meta.db_table,
+                "column": field.column,
```
Table and column names must be quoted.

felixxm commented on docs/ref/schema-editor.txt:
```diff
@@ -127,6 +127,16 @@ value.
 
+``alter_db_table_comment()``
+----------------------------
+
```
`versionadded` annotation is missing.

KimSoungRyoul commented:
buildbot, test on oracle.

ngnpope commented on django/db/backends/postgresql/schema.py:
```diff
@@ -58,6 +58,44 @@ def quote_value(self, value):
         # getquoted() returns a quoted bytestring of the adapted value.
         return adapted.getquoted().decode()
 
+    def quote_comment_value(self, value):
+        return self.quote_value(value)
+
```
```suggestion
```

It turns out that we don't need to have specific behaviour for PostgreSQL which treats `''` and `NULL` the same.

See [this fiddle](https://dbfiddle.uk/8SiuoziM) for details.

KimSoungRyoul commented:
buildbot, test on oracle.

ngnpope commented:
Hi @KimSoungRyoul. Thanks for sticking with this. I've created https://github.com/Ice-Caramel-Macchiato/django/pull/2 which has the changes I proposed about adding a `supports_comments_inline` feature flag. Please could you test this out? It has the advantage that third-party backends may not need to override `create_model()` or `add_field()` - in particular there are a lot of PostgreSQL-compatible implementations, e.g. CockroachDB, that would have to make those changes as things stand.

I was also looking at some other questions:

- Should support comments on other objects, e.g. database, index, constraint, etc?
- Should we add system checks for comment length to prevent avoidable crashes when running migrations?
- Should we split `supports_comments` into e.g. `supports_comment_on_table`, `supports_comment_on_column`, etc?

```
                Database    Table   Column  Index   Constraint  Sequence    Max Length
PostgreSQL      Y           Y       Y       Y       Y           Y           Essentially unlimited.
MariaDB         >=10.5      Y       Y       Y       Y           Y           2048 if table else 1024.
MySQL           N           Y       Y       Y       Y           Y           2048 if table else 1024.
SQLite          N           N       N       N       N           N           N/A
Oracle          N           Y       Y       Y       N           N           4000
``` 

Regarding checking the length of comments - either this could be managed again by a feature flag for each backend, but it may be better to just limit to 1024 characters across all backends to avoid issues where multiple backends may be used?

We could add support for comments on indexes under the same flag on the basis of the above, but for constraints, we'd then probably want to consider splitting the flags based on the type of object... 🤔

Although comments on the database itself are not widely supported, is this something we may still want? It'd have to be something set in the `DATABASES` setting. It could also be wired up when creating databases in tests to add a comment that it is a temporary database created for testing by Django?

What are your thoughts @carltongibson and @felixxm?

ngnpope commented on tests/schema/tests.py:
```diff
@@ -273,6 +278,16 @@ def get_column_collation(self, table, column):
                 if f.name == column
             )
 
+    def get_column_comment(self, table, column):
+        with connection.cursor() as cursor:
+            columns = connection.introspection.get_table_description(cursor, table)
+        return [c.comment for c in columns if c.name == column][0]
+
+    def assert_column_comment_not_exists(self, table, column):
+        with connection.cursor() as cursor:
+            columns = connection.introspection.get_table_description(cursor, table)
+        self.assertFalse(any([c.name == column and c.comment for c in columns]))
+
```
Perhaps this could be aligned with the similar implementation of these in `MigrationTestBase`? Perhaps extract into a mixin that both could use? Lots of `assertEqual()` here could be replaced with `assertColumnCommentEquals()`?

KimSoungRyoul commented:
> Hi @KimSoungRyoul. Thanks for sticking with this. I've created [Ice-Caramel-Macchiato#2](https://github.com/Ice-Caramel-Macchiato/django/pull/2) which has the changes I proposed about adding a `supports_comments_inline` feature flag. Please could you test this out? It has the advantage that third-party backends may not need to override `create_model()` or `add_field()` - in particular there are a lot of PostgreSQL-compatible implementations, e.g. CockroachDB, that would have to make those changes as things stand.
> 
> I was also looking at some other questions:
> 
> * Should support comments on other objects, e.g. database, index, constraint, etc?
> * Should we add system checks for comment length to prevent avoidable crashes when running migrations?
> * Should we split `supports_comments` into e.g. `supports_comment_on_table`, `supports_comment_on_column`, etc?
> 
> ```
>                 Database    Table   Column  Index   Constraint  Sequence    Max Length
> PostgreSQL      Y           Y       Y       Y       Y           Y           Essentially unlimited.
> MariaDB         >=10.5      Y       Y       Y       Y           Y           2048 if table else 1024.
> MySQL           N           Y       Y       Y       Y           Y           2048 if table else 1024.
> SQLite          N           N       N       N       N           N           N/A
> Oracle          N           Y       Y       Y       N           N           4000
> ```
> 
> Regarding checking the length of comments - either this could be managed again by a feature flag for each backend, but it may be better to just limit to 1024 characters across all backends to avoid issues where multiple backends may be used?
> 
> We could add support for comments on indexes under the same flag on the basis of the above, but for constraints, we'd then probably want to consider splitting the flags based on the type of object... 🤔
> 
> Although comments on the database itself are not widely supported, is this something we may still want? It'd have to be something set in the `DATABASES` setting. It could also be wired up when creating databases in tests to add a comment that it is a temporary database created for testing by Django?
> 
> What are your thoughts @carltongibson and @felixxm?

* #### Should support comments on other objects, e.g. database, index, constraint, etc?
no I think that it's considered an unnecessary feature. I'm not sure if I should support it this deeply
and also other ORM frameworks don't support like this


* #### Should we add system checks for comment length to prevent avoidable crashes when running migrations?
I want raising error on database Layer. It doesn't seem bad to give a warning at the time of `execute makemigrations`


* #### Should we split supports_comments into e.g. supports_comment_on_table, supports_comment_on_column, etc?
It seems to be an overly detailed feature detail.  feature(`supports_comment_on_table`, `supports_comment_on_column`) seems to be sufficient to improve when django supports the other database 

I hope you don't feel my writing sharp because of my poor English writing skills
thank you for your feedback 😄

KimSoungRyoul commented on docs/ref/models/options.txt:
```diff
@@ -101,6 +101,33 @@ Django quotes column and table names behind the scenes.
+                           the user registers. This comment is managed by\
+                           Django, and will be updated on every migration."
+
+``db_table_comment`` has a character limit that differs for each database
+backend. PostgreSQL has a 1GB limit. MySQL has a 2048 character limit.
```
you are right , It is appropriate that this sentence be deleted from the document

KimSoungRyoul commented:
buildbot, test on oracle.

felixxm commented:
@KimSoungRyoul I've started to work on this patch. Can you allow edits by maintainers? I'm checking all comments and would like to push small edits:

![image](https://user-images.githubusercontent.com/2865885/208400781-04118b70-bf67-4775-860c-6f3546853869.png)

KimSoungRyoul commented:
@felixxm 
`Ice-Caramel-Macchiato:support-migrate-with-db-comment` Repo owner is Organization 
I'm not sure if this is the cause of the problem, there is no checkbox(`Allow edits and access~`) so I sent you an email to give you permission 
<img width="345" alt="스크린샷 2022-12-19 오후 8 32 10" src="https://user-images.githubusercontent.com/24240623/208416734-8385884b-1ef6-4a32-8a3d-ac983aafa20b.png">

Now you should have permission to commit ,push 
check your email box

felixxm commented:
Thanks :+1: Rebased, I'm checking comments ...

felixxm commented:
buildbot, test on oracle.

felixxm commented:
> * Should support comments on other objects, e.g. database, index, constraint, etc?

No, at least in this PR/ticket.

> * Should we add system checks for comment length to prevent avoidable crashes when running migrations?

The length of comments is not particularly small, at least 1023 (MySQL), so as for me database error is fine when someone hits the limit.
 
> * Should we split `supports_comments` into e.g. `supports_comment_on_table`, `supports_comment_on_column`, etc?

A single database feature flag (`supports_comments`) should be sufficient.

rsalmaso commented:
Seen this PR via @felixxm toot :smile:
It seems that this PR doesn't allow to use the verbose_name as comment like [django-db-comments](https://pypi.org/project/django-db-comments/) does (which we use at $job).
Can be `db_comment=True` solve this?
Or maybe better an Meta option to apply to all fields?

FYI `django-db-comments` adds comments for all models, which is fancy, maybe an helper method called in project `AppConfig.ready()` could be fine? :thinking:

felixxm commented:
> Seen this PR via @felixxm toot smile It seems that this PR doesn't allow to use the verbose_name as comment like [django-db-comments](https://pypi.org/project/django-db-comments/) does (which we use at $WORK). Can be `db_comment=True` solve this? Or maybe better an Meta option to apply to all fields?
> FYI `django-db-comments` adds comments for all models, which is fancy, maybe an helper method called in project `AppConfig.ready()` could be fine? thinking

As far as I'm aware, basing `db_comment` on `verbose_name` won't work for most of projects, it's too generic. Also, `verbose_name`s are for users and `db_comment`s are for DBAs. You can still use `django-db-comments` if it works for you, that's why we have 3rd-party packages :package:

rsalmaso commented:
> As far as I'm aware, basing `db_comment` on `verbose_name` won't work for most of projects, it's too generic. Also, `verbose_name`s are for users and `db_comment`s are for DBAs. You can still use `django-db-comments` if it works for you, that's why we have 3rd-party packages package

Fine, make sense.
I would be happy to remove a dependancy if it is possible :smile_cat:

felixxm commented on django/db/migrations/autodetector.py:
```diff
@@ -1552,6 +1553,28 @@ def generate_altered_db_table(self):
                     ),
                 )
 
+    def generate_altered_db_table_comment(self):
+        models_to_check = self.kept_model_keys.union(
+            self.kept_proxy_keys, self.kept_unmanaged_keys
+        )
+        for app_label, model_name in sorted(models_to_check):
+            old_model_name = self.renamed_models.get(
+                (app_label, model_name), model_name
+            )
+            old_model_state = self.from_state.models[app_label, old_model_name]
+            new_model_state = self.to_state.models[app_label, model_name]
+
+            old_db_table_comment = old_model_state.options.get("db_table_comment")
+            new_db_table_comment = new_model_state.options.get("db_table_comment")
+            if old_db_table_comment != new_db_table_comment:
+                self.add_operation(
+                    app_label,
+                    operations.AlterModelTableComment(
+                        name=model_name,
+                        db_table_comment=new_db_table_comment,
+                    ),
+                )
+
```
This is untested.

felixxm replied:
Added tests.

felixxm commented on django/db/models/base.py:
```diff
@@ -1617,6 +1617,14 @@ def _check_swappable(cls):
+                    obj=cls,
+                    id="models.W045",
+                )
+            )
```
This should be moved to the separate method and take `databases` and `required_db_features` into account.

felixxm replied:
This warning is also untested.

felixxm replied:
Tests added.

felixxm commented on django/db/models/base.py:
```diff
@@ -1617,6 +1617,14 @@ def _check_swappable(cls):
+                checks.Warning(
+                    f"{connection.display_name} does not support db_table_comment.",
+                    obj=cls,
+                    id="models.W045",
```
`models.W045` is already used:
```suggestion
                    id="models.W046",
```

felixxm commented on tests/invalid_models_tests/test_relative_fields.py:
```diff
@@ -89,6 +89,36 @@ class Model(models.Model):
         field = Model._meta.get_field("m2m")
         self.assertEqual(field.check(from_model=Model), [])
 
+    def test_many_to_many_with_limit_choices_auto_created_no_warning(self):
+        class Model(models.Model):
+            name = models.CharField(max_length=20)
+
+        class ModelM2M(models.Model):
+            m2m = models.ManyToManyField(Model, limit_choices_to={"name": "test_name"})
+
+        self.assertEqual(ModelM2M.check(), [])
```
This test is unnecessary, other tests will fail if we raise a warning for `ManyToManyField` without `db_comment`.

felixxm commented on tests/introspection/tests.py:
```diff
@@ -79,6 +80,11 @@ def test_unmanaged_through_model(self):
         tables = connection.introspection.django_table_names()
         self.assertNotIn(ArticleReporter._meta.db_table, tables)
 
+    @skipUnlessDBFeature("supports_comments", "supports_foreign_keys")
+    def test_unmanaged_db_comment_model(self):
+        tables = connection.introspection.django_table_names()
+        self.assertNotIn(ArticleReporterWithDBComment._meta.db_table, tables)
```
I'm not sure why adding `db_table_comment` would change Django's behavior for unmanaged models :thinking: This looks unnecessary.

felixxm replied:
Removed.

felixxm reviewed (commented):
Added introspection of table comments.

felixxm commented:
Added `inspectdb` support with tests.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
I am 65% satisfied with this patch :chart_with_upwards_trend:, so we're making progress :clock1:.
I'm going to continue next week :santa:, mostly on docs and tests :factory_worker:

felixxm reviewed (commented):
I pushed edits to tests.

felixxm commented:
@KimSoungRyoul Thanks for all your efforts :1st_place_medal: Welcome aboard :boat: 

Thanks y'all for reviews :star2: 

It's ready :rocket:

felixxm commented:
buildbot, test on oracle.

KimSoungRyoul commented:
Thanks to your reviews, the pullrequest was completed well. 😄

timgraham commented on django/db/backends/postgresql/introspection.py:
```diff
@@ -62,7 +63,8 @@ def get_table_list(self, cursor):
                     ELSE 't'
-                END
+                END,
+                obj_description(c.oid)
```
While looking at CockroachDB support, the following from [the PostgreSQL documentation](https://www.postgresql.org/docs/15/functions-info.html#FUNCTIONS-INFO-COMMENT-TABLE) was brought to my attention:
```
obj_description ( object oid ) → text

Returns the comment for a database object specified by its OID alone.
This is deprecated since there is no guarantee that OIDs are unique
across different system catalogs; therefore, the wrong comment might
be returned.
```

felixxm replied:
Good catch :dart: We should add a catalog name.
